---
title: "sambadaOnR"
author: "Solange Gaillard, Sylvie Stucki, Oliver Selmoni, Elia Vajana"
date: "`r Sys.Date()`"
output: rmarkdown::html_document
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This package is designed to entail the whole processing chain to use SamBada, from pre- to post-processing. In this documentation, we will work through all these steps with an example given in the data available with the package. This is a subset of ten SNPs from 800 Ougandan cattle including the sample location (see SamBada's documentation below for more information)

If you want more documentation, you can read the [documentation of the package](https://github.com/SolangeD/sambadaOnR/sambadaOnR-manual.pdf) or [SamBada's documentation](https://github.com/Sylvie/sambada/releases/download/v0.7.1/sambadoc.pdf). Also read the article ...

## Download Sambada

For running sambada, you need to download sambada's binaries. This can be done with `downloadSambada` which downloads Sambada from GitHub and upacks it into the directory of your choice. You might already be a Sambada user and do not have to download it again! Note that if you plan to often use Sambada, it is recommended that you put the binaries folder of Sambada into your path environmental variable (this procedure is OS-dependent, look on the internet how to proceed), otherwise you will have to specify this path every time you start a new R session.

```{r, eval=FALSE}
#Load help 
?downloadSambada()
#Downloads Sambada into the current active directory
downloadSambada()
```

## Preprocessing

Most of the functions described here have an `interactiveChecks` mode. When you run the function for the first time on your dataset, we strongly advise that you set it to TRUE. This prints plots that allows you to detect anomalies.

### Prepare the genomic file

The first step when you have your genomic matrix is to prepare it into a format that samBada accepts. You can use `prepareGeno` for this, which can process plink ped, plink bed, vcf or gds input file. In the meantime, you can also filter out SNPs based on Minor Allele Frequency (MAF), Missingness, Linkage Disequilibrium (LD) and Major Genotype Frequency (MGF). 

```{r, eval=FALSE}
#Loads data
#================
#These files are distributed within your package. The system.file will return the full path to them. With your data, you can just use the name of the file, provided the file is in the active directory
genoFile=system.file("extdata", "uganda-subset.ped", package = "sambadaOnR")
genoFile #Check the path to the file

#Load help
#================
?prepareGeno

#Run prepareGeno
#================
#Warning: the histograms shown due to maf and missingness filtering are really difficult to understand given the small number of SNPs
prepareGeno(fileName=genoFile,mafThresh=0.05, missingnessThresh=0.1,interactiveChecks=TRUE)
```

### Create the environmental dataset

Then from the point locations, you need to create your envionmental dataset from point location. Use `createEnv` for this task.  

You can use rasters of your study site that you already have or use the function to automatically download rasters of your study site from global databases [see worldclim dataset]() for information on the downloaded tiles. For example tminXX represents the minimum temperature in october. Similarly tmax, tavg and prec refers to maximum temperature, average temperature and precipitation. The bio1-bio19 are bioclim variables are computed from these indices and are described [here](www.worldclim.org/current). Temperature are given in 10 degree C and precipitation in mm. The funciton always downloads the best resolution available (30 seconds for worldclim dataset and 90m for SRTM). 

The tiles are downloaded in the folders wc0.5 and srtm of the active directory. The downloaded tiles can get really heavy! Once your final file is exported, you could delete the tiles (but you won't be able to use them as background in the post-processing mapping function). If the tiles are already present in the active directory, they won't be downloaded again. 

This function requires that you define the EPSG code of your projection system. All systems are referenced [here](http://spatialreference.org/ref/epsg/?search=4326&srtext=Search) If you work with lat/long global projection, then you most probably work with WGS 84 whose EPSG is 4326.
```{r, eval=FALSE}
#Loads data
#================
#These files are distributed within your package. The system.file will return the full path to them. With your data, you can just use the name of the file, provided the file is in the active directory
locationFile=system.file("extdata", "uganda-subset.csv", package = "sambadaOnR")
locationFile #Check the path to the file

#Load help
#================
?createEnv

#Create environmental dataset
#================
#downloads the raster tiles from global databases and create the environmental file
#Warning: the download and processing of raster is both heavy in space and time-consuming
#If you want to skip this step, you can directly use the file in ...
createEnv(locationFileName=locationFile, x='longitude',y='latitude',locationProj=4326,separator=';', rasterName=NULL,rasterProj=NULL, worldclim=TRUE, interactiveChecks=TRUE)

```

### Prepare the final environmental dataset

You can now use the `prepareEnv` function. This function has 3 goals

* Put the sample ID of the genomic file and the environmental file in the same order (required to run sambada)  
* Reduce your environmental dataset. Indeed, if you use worlclim variables for examples, some of the variables will be very correlated. We can delete correlated variables that are above a given correlation threshold (argument `maxCorr`)  
* Check if there is a population structure to include it as an "environmental variable" in your environmental file.

The function creates a new file with the '-export'-suffix at the end of the envFile.

The population structure is calculated as a PCA of all the SNPs that pass the filtering (maf, ld, missingness). You can either choose to use the score of the X first components to evaluate the population structure (set `numPop` to NULL) or you can compute a "membership coefficient" to a cluster of individuals based on the scores on the first X components. You can choose between two clustering algorithm (k-means or hierarchical cluster in the `clustMethod` argument).

One of the option to decide the number of PCs that you should keep is to detect a bump in the proportion of variance explained and keep all the PC before the bump.

```{r, eval=FALSE}
#Loads data
#================
#If you ran prepareGeno, use the gds file created in this step 
gdsFile='uganda-subset.gds'
#Otherwise, take the one prvided in the package
gdsFile=system.file("extdata", "uganda-subset.gds", package = "sambadaOnR")
gdsFile #Check the path to the file
#If you ran createEnv, use the .csv file created in this step
envFile='uganda-subset-env.csv'
#Otherwise, take the one prvided in the package
envFile=system.file("extdata", "uganda-subset-env.csv", package = "sambadaOnR")
envFile #Check the path to the file

#Load help
#================
?prepareEnv

#prepareEnv
#================

#Stores Principal components scores
prepareEnv(envFile=envFile, maxCorr=0.8, idName='short_name', genoFile=gdsFile, numPc=0.2, mafThresh=0.05, missingnessThresh=0.1, ldThresh=0.2, numPop=NULL, x='longitude', y='latitude', interactiveChecks=TRUE, locationProj=4326 )
#Accept the maxCorr threshold, but set the numPc to 1.

```

## Run SamBada

If you run samBada from the command line, you will have to create a parameter file. All parameters that can be calculated automatically from your file are calculated for you in `sambadaParallel`. 

Furthermore, samBada includes a module called supervision to split the molecular file into several subfiles to allow parallel computing. The split of the input file and merge of the output file with supervision's help is integrated in the function.

```{r, eval=FALSE}
#Loads data
#================
#If you ran prepareEnv, use the .csv file created in this step
envFile='uganda-subset-env-export.csv'
#Otherwise, take the one prvided in the package
envFile=system.file("extdata", "uganda-subset-env-export.csv", package = "sambadaOnR")
envFile #Check the path to the file
#If you ran prepareGeno, use the .csv file created in this step
genoFile='uganda-subset.csv'
#Otherwise, take the one prvided in the package
genoFile=system.file("extdata", "uganda-subset.csv", package = "sambadaOnR")
genoFile #Check the path to the file

#Load help
#================
?sambadaParallel

#sambadaParallel
#================
#Run sambada on two cores. 
#prepareEnv puts the population variables at the end of the file (=> LAST). 
#Only one pop var was saved, so set dimMax=2 
sambadaParallel(genoFile=genoFile, envFile=envFile, idGeno='ID_indiv', idEnv='short_name', dimMax=2, cores=2, saveType='END ALL', populationVar='LAST')

```
## Post-processing

### Prepare the output

The function `prepareOutput` does the following things on sambada's output

* Calculate p and q-value
* Retrieves the chromosome and position of each SNP in the output from the GDS file
* Filter out models that are not interesting (if you ran a multivariate model with population structure)
* It also computes the maximum position in each chromosome to ease the plotting of manhattan plot.

```{r, eval=FALSE}
#Loads data
#================
#If you haven't run sambadaParallel, you need to copy the output file into the active directory with the following command
file.copy(system.file("extdata", "uganda-subset-Out-2.csv", package = "sambadaOnR"), 'uganda-subset-Out-2.csv')
#Furthermore, if you haven't run prepareGeno, you need to copy the GDS file with the following command
gdsFile=file.copy(system.file("extdata", "uganda-subset.gds", package = "sambadaOnR"),'uganda-subset.gds')


#Load help
#================
?prepareOutput

#prepareOutput
#================
prep = prepareOutput(sambadaname='uganda-subset', dimMax=2, popStr=TRUE, interactiveChecks=TRUE)

```
### Manhattan plots

To explore your results, the first thing you could do is draw a manhattan plot of each environmental variable to detect one or several intersting peaks in particular variables

```{r, eval=FALSE}
#Loads data
#================
#You need to run prepareOutput to run this function

#Load help
#================
?plotManhattan

#plotManhattan
#================
#Plot manhattan of all kept variables
plotManhattan(prep, c('bio1' 'bio2' 'bio3' 'bio6' 'bio12' 'bio14' 'bio15' 'bio19' 'prec4' 'prec5' 'prec7' 'prec9' 'prec10'),chromo='all',valueName='pvalueG')
```


### Plot the results interactively

The function `plotResultInteractive` can now be invoked. This will open a local page on your web-browser with a manhattan plot (you have to choose the environmental variable you want to study and can choose the chromosomes to draw). 

The plot is interactive so that you can select a point on the manhattan which will query the [Ensembl database](http://www.ensembl.org/index.html) to indicate nearby SNPs and the consequence of the variant (intergenic, synonymous, non-synonymous, stop gained, stop lost,...) with [VEP](www.ensembl.org/info/docs/tools/vep/index.html). 

The map of the marker, population variables and environmental variable is available. 

A boxplot showing the environmental range of both present and absent individuals is drawn as well.

```{r, eval=FALSE}
#Loads data
#================
#You need to run prepareOutput to run this function
#If you haven't run prepareEnv, use the file provided in the package
envFile=system.file("extdata", "uganda-subset-env-export.csv", package = "sambadaOnR")
#Otherwise use
envFile='uganda-subset-env-export.csv'
#If you haven't run prepareGeno, use the file provided in the package
gdsFile=system.file("extdata", "uganda-subset.gds", package = "sambadaOnR")
#Otherwise use
gdsFile=NULL #will guess it from envFile name

#Load help
#================
?plotResultInteractive

#plotResultInteractive
#================

plotResultInteractive(preparedOutput=prep, varEnv='bio1', envFile=envFile,species='btaurus', pass=50000,x='longitude',y='latitude',gdsFile=gdsFile, IDCol='short_name', popstrCol='pop1')
```

### Plot maps

You might be interested in plotting a map (though it is already done in the `plotResultInteractive`). The advantages of the `plotMap` are

* Use of raster as background if available
* Closely located points are scattered so that all points are visible (or should be!)

You can draw a map of

* marker distribution (presence/absence)
* population structure (either as pie charts if membership coefficient or as a continuous variable)
* envrionmental variable (if no raster available)
* Auto-correlation of marker

```{r, eval=FALSE}
#Loads data
#================
#You need to run prepareOutput to run this function
#If you haven't run prepareEnv, use the file provided in the package
envFile=system.file("extdata", "uganda-subset-env-export.csv", package = "sambadaOnR")
#Otherwise use
envFile='uganda-subset-env-export.csv'
#If you haven't run prepareGeno, use the file provided in the package
gdsFile=system.file("extdata", "uganda-subset.gds", package = "sambadaOnR")
#Otherwise use
gdsFile=NULL #will guess it from envFile name

#Load help
#================
?plotMap

#plotMap
#================

plotMap = function(envFile=envFile, x='longitude', y='latitude', locationProj=4326,  popStrCol='pop1', gdsFile=gdsFile, markerName, mapType='marker', varEnvName='bio1', simultaneous=FALSE){

```

Good luck with the analysis of your own dataset!

<!-- ## Figures -->

<!-- The figure sizes have been customised so that you can easily put two images side-by-side.  -->

<!-- ```{r, fig.show='hold'} -->
<!-- plot(1:10) -->
<!-- plot(10:1) -->
<!-- ``` -->

<!-- You can enable figure captions by `fig_caption: yes` in YAML: -->

<!--     output: -->
<!--       rmarkdown::html_vignette: -->
<!--         fig_caption: yes -->

<!-- Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**. -->

<!-- ## More Examples -->

<!-- You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`. -->

<!-- ```{r, echo=FALSE, results='asis'} -->
<!-- knitr::kable(head(mtcars, 10)) -->
<!-- ``` -->

<!-- Also a quote using `>`: -->

<!-- > "He who gives up [code] safety for [code] speed deserves neither." -->
<!-- ([via](https://twitter.com/hadleywickham/status/504368538874703872)) -->
